File: src/src_core/fmg/core/img/MosaicGroupTransformer.cs
using System;
using fmg.core.types;

namespace fmg.core.img {

    /// <summary> Transforming of <see cref="MosaicGroupModel"/> </summary>
    public class MosaicGroupTransformer : IModelTransformer {

        public void Execute(IAnimatedModel model) {
            if (!(model is MosaicGroupModel m))
                throw new InvalidOperationException("Illegal usage transformer");

            if (MosaicGroupModel.varMosaicGroupAsValueOthers1 && (m.MosaicGroup == EMosaicGroup.eOthers)) {
                bool castling = false;
                double rotateAngleDelta = 360.0 / m.TotalFrames; // 360° / TotalFrames
                if (!m.AnimeDirection)
                    rotateAngleDelta = -rotateAngleDelta;
                double incrementSpeedAngle = m.IncrementSpeedAngle + 3 * rotateAngleDelta;
                if (incrementSpeedAngle >= 360) {
                    incrementSpeedAngle -= 360;
                    castling = true;
                } else {
                    if (incrementSpeedAngle < 0) {
                        incrementSpeedAngle += 360;
                        castling = true;
                    }
                }
                m.IncrementSpeedAngle = incrementSpeedAngle;
                if (castling) {
                    m.NmIndex1 = (m.NmIndex1 + 1) % m.NmArray.Length;
                    m.NmIndex2 = (m.NmIndex2 + 1) % m.NmArray.Length;
                }
            }
        }

    }

}

File: src/src_core/fmg/core/img/MosaicGroupTransformer.java
package fmg.core.img;

import fmg.core.types.EMosaicGroup;

/** Transforming of {@link MosaicGroupModel} */
public class MosaicGroupTransformer implements IModelTransformer {

   @Override
   public void execute(IAnimatedModel model) {
      if (!(model instanceof MosaicGroupModel))
         throw new RuntimeException("Illegal usage transformer");

      MosaicGroupModel m = (MosaicGroupModel)model;

      if (MosaicGroupModel.varMosaicGroupAsValueOthers1 && (m.getMosaicGroup() == EMosaicGroup.eOthers)) {
         boolean castling = false;
         double rotateAngleDelta = 360.0 / m.getTotalFrames(); // 360° / TotalFrames
         if (!m.getAnimeDirection())
            rotateAngleDelta = -rotateAngleDelta;
         double incrementSpeedAngle = m.getIncrementSpeedAngle() + 3*rotateAngleDelta;
         if (incrementSpeedAngle >= 360) {
            incrementSpeedAngle -= 360;
            castling = true;
         } else {
            if (incrementSpeedAngle < 0) {
               incrementSpeedAngle += 360;
               castling = true;
            }
         }
         m.setIncrementSpeedAngle(incrementSpeedAngle);
         if (castling) {
            m.setNmIndex1((m.getNmIndex1()+1) % m.getNmArray().length);
            m.setNmIndex2((m.getNmIndex2()+1) % m.getNmArray().length);
         }
      }
   }

}
